// typebox-validators@0.3.5/heterogeneous/heterogeneous-union-validator.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _HeterogeneousUnionValidator_typeIdentifyingKeyIndex;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeterogeneousUnionValidator = void 0;
const value_1 = require("@sinclair/typebox/value");
const abstract_typed_union_validator_1 = require("../abstract/abstract-typed-union-validator");
const error_utils_1 = require("../lib/error-utils");
const type_identifying_key_index_1 = require("./type-identifying-key-index");
/**
 * Non-compiling validator for heterogeneous unions of objects. To improve
 * performance, list the more frequently used types earlier in the union, and
 * list each object's unique key first in its properties.
 */
class HeterogeneousUnionValidator extends abstract_typed_union_validator_1.AbstractTypedUnionValidator {
    /** @inheritdoc */
    constructor(schema) {
        super(schema);
        _HeterogeneousUnionValidator_typeIdentifyingKeyIndex.set(this, void 0);
        __classPrivateFieldSet(this, _HeterogeneousUnionValidator_typeIdentifyingKeyIndex, new type_identifying_key_index_1.TypeIdentifyingKeyIndex(schema), "f");
    }
    /** @inheritdoc */
    test(value) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            return false;
        }
        return value_1.Value.Check(this.schema.anyOf[indexOrError], value);
    }
    /** @inheritdoc */
    errors(value) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            return (0, error_utils_1.createUnionTypeErrorIterable)(indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        return (0, error_utils_1.createErrorsIterable)(value_1.Value.Errors(schema, value));
    }
    assertReturningSchema(value, overallError) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidAssert)(overallError, indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        this.uncompiledAssert(schema, value, overallError);
        return schema;
    }
    validateReturningSchema(value, overallError) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidValidate)(overallError, indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        this.uncompiledValidate(schema, value, overallError);
        return schema;
    }
    findSchemaMemberIndex(value) {
        if (__classPrivateFieldGet(this, _HeterogeneousUnionValidator_typeIdentifyingKeyIndex, "f").keyByMemberIndex === undefined) {
            // only incur cost if validator is actually used
            __classPrivateFieldGet(this, _HeterogeneousUnionValidator_typeIdentifyingKeyIndex, "f").cacheKeys();
        }
        if (typeof value === 'object' && value !== null) {
            for (let i = 0; i < this.schema.anyOf.length; ++i) {
                const uniqueKey = __classPrivateFieldGet(this, _HeterogeneousUnionValidator_typeIdentifyingKeyIndex, "f").keyByMemberIndex[i];
                if (value[uniqueKey] !== undefined) {
                    return i;
                }
            }
        }
        return (0, error_utils_1.createUnionTypeError)(this.schema, value);
    }
}
exports.HeterogeneousUnionValidator = HeterogeneousUnionValidator;
_HeterogeneousUnionValidator_typeIdentifyingKeyIndex = new WeakMap();
