// typebox-validators@0.3.5/heterogeneous/type-identifying-key-index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeIdentifyingKeyIndex = exports.MESSAGE_MEMBERS_MISSING_KEY = exports.MESSAGE_MULTIPLE_MEMBERS_WITH_SAME_KEY = exports.MESSAGE_MEMBER_WITH_MULTIPLE_KEYS = exports.MESSAGE_OPTIONAL_TYPE_ID_KEY = void 0;
const typebox_1 = require("@sinclair/typebox");
exports.MESSAGE_OPTIONAL_TYPE_ID_KEY = 'Type identifying key cannot be optional';
exports.MESSAGE_MEMBER_WITH_MULTIPLE_KEYS = 'Union has member with multiple identifying keys';
exports.MESSAGE_MULTIPLE_MEMBERS_WITH_SAME_KEY = 'Union has multiple members with same identifying key';
exports.MESSAGE_MEMBERS_MISSING_KEY = 'Union has members missing identifying keys';
// Note: The type identifying keys of heterogeneous unions are assigned via
// the typeIdentifyingKey schema option instead of being inferred, because
// inferrence requires a lot of computation, especially because multiple
// keys might be unique within a given member schema.
class TypeIdentifyingKeyIndex {
    constructor(schema) {
        this.schema = schema;
    }
    cacheKeys() {
        const unionSize = this.schema.anyOf.length;
        const takenKeys = new Set();
        this.keyByMemberIndex = new Array(unionSize);
        for (let i = 0; i < unionSize; ++i) {
            const memberSchema = this.schema.anyOf[i];
            for (const [key, schema] of Object.entries(memberSchema.properties)) {
                if (schema.typeIdentifyingKey) {
                    if (schema[typebox_1.Optional] == 'Optional') {
                        throw Error(exports.MESSAGE_OPTIONAL_TYPE_ID_KEY);
                    }
                    if (this.keyByMemberIndex[i] !== undefined) {
                        throw Error(exports.MESSAGE_MEMBER_WITH_MULTIPLE_KEYS);
                    }
                    if (takenKeys.has(key)) {
                        throw Error(exports.MESSAGE_MULTIPLE_MEMBERS_WITH_SAME_KEY);
                    }
                    this.keyByMemberIndex[i] = key;
                    takenKeys.add(key);
                }
            }
        }
        if (takenKeys.size < unionSize) {
            this.keyByMemberIndex = undefined; // reset for next attempt
            throw Error(exports.MESSAGE_MEMBERS_MISSING_KEY);
        }
    }
}
exports.TypeIdentifyingKeyIndex = TypeIdentifyingKeyIndex;
