// typebox-validators@0.3.5/discriminated/discriminated-union-validator.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _DiscriminatedUnionValidator_unionIsWellformed;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscriminatedUnionValidator = void 0;
const value_1 = require("@sinclair/typebox/value");
const abstract_typed_union_validator_1 = require("../abstract/abstract-typed-union-validator");
const error_utils_1 = require("../lib/error-utils");
/**
 * Non-compiling validator for discriminated unions. To improve performance,
 * list the more frequently used types earlier in the union, and list each
 * object's discriminant key first in its properties.
 */
class DiscriminatedUnionValidator extends abstract_typed_union_validator_1.AbstractTypedUnionValidator {
    /** @inheritdoc */
    constructor(schema) {
        var _a;
        super(schema);
        _DiscriminatedUnionValidator_unionIsWellformed.set(this, false);
        this.discriminantKey =
            (_a = this.schema.discriminantKey) !== null && _a !== void 0 ? _a : abstract_typed_union_validator_1.DEFAULT_DISCRIMINANT_KEY;
    }
    /** @inheritdoc */
    test(value) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            return false;
        }
        return value_1.Value.Check(this.schema.anyOf[indexOrError], value);
    }
    /** @inheritdoc */
    errors(value) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            return (0, error_utils_1.createUnionTypeErrorIterable)(indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        return (0, error_utils_1.createErrorsIterable)(value_1.Value.Errors(schema, value));
    }
    assertReturningSchema(value, overallError) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidAssert)(overallError, indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        this.uncompiledAssert(schema, value, overallError);
        return schema;
    }
    validateReturningSchema(value, overallError) {
        const indexOrError = this.findSchemaMemberIndex(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidValidate)(overallError, indexOrError);
        }
        const schema = this.schema.anyOf[indexOrError];
        this.uncompiledValidate(schema, value, overallError);
        return schema;
    }
    findSchemaMemberIndex(subject) {
        if (!__classPrivateFieldGet(this, _DiscriminatedUnionValidator_unionIsWellformed, "f")) {
            // only incur cost if validator is actually used
            for (const memberSchema of this.schema.anyOf) {
                if (memberSchema.properties[this.discriminantKey] === undefined) {
                    throw Error(`Discriminant key '${this.discriminantKey}' not present in all members of discriminated union`);
                }
            }
            __classPrivateFieldSet(this, _DiscriminatedUnionValidator_unionIsWellformed, true, "f");
        }
        if (typeof subject === 'object' && subject !== null) {
            const subjectKind = subject[this.discriminantKey];
            if (subjectKind !== undefined) {
                for (let i = 0; i < this.schema.anyOf.length; ++i) {
                    const memberKind = this.schema.anyOf[i].properties[this.discriminantKey];
                    if (memberKind !== undefined && memberKind.const === subjectKind) {
                        return i;
                    }
                }
            }
        }
        return (0, error_utils_1.createUnionTypeError)(this.schema, subject);
    }
}
exports.DiscriminatedUnionValidator = DiscriminatedUnionValidator;
_DiscriminatedUnionValidator_unionIsWellformed = new WeakMap();
