// snowflakify@1.0.6/dist/cjs/index.js
"use strict";
// Copyright (c) Ariana (CatWithAWand). All rights reserved. Licensed under the MIT license.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceFragment = exports.RandomFragment = exports.NetworkFragment = exports.ProcessFragment = exports.WorkerFragment = exports.TimestampFragment = exports.default = void 0;
/**
 * A module for generating snowflake IDs.
 *
 * @packageDocumentation
 */
var Snowflakify_js_1 = require("./Snowflakify.js");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(Snowflakify_js_1).default; } });
var TimestampFragment_js_1 = require("./Fragments/TimestampFragment.js");
Object.defineProperty(exports, "TimestampFragment", { enumerable: true, get: function () { return __importDefault(TimestampFragment_js_1).default; } });
var WorkerFragment_js_1 = require("./Fragments/WorkerFragment.js");
Object.defineProperty(exports, "WorkerFragment", { enumerable: true, get: function () { return __importDefault(WorkerFragment_js_1).default; } });
var ProcessFragment_js_1 = require("./Fragments/ProcessFragment.js");
Object.defineProperty(exports, "ProcessFragment", { enumerable: true, get: function () { return __importDefault(ProcessFragment_js_1).default; } });
var NetworkFragment_js_1 = require("./Fragments/NetworkFragment.js");
Object.defineProperty(exports, "NetworkFragment", { enumerable: true, get: function () { return __importDefault(NetworkFragment_js_1).default; } });
var RandomFragment_js_1 = require("./Fragments/RandomFragment.js");
Object.defineProperty(exports, "RandomFragment", { enumerable: true, get: function () { return __importDefault(RandomFragment_js_1).default; } });
var SequenceFragment_js_1 = require("./Fragments/SequenceFragment.js");
Object.defineProperty(exports, "SequenceFragment", { enumerable: true, get: function () { return __importDefault(SequenceFragment_js_1).default; } });
