// typebox-validators@0.3.5/lib/error-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwInvalidValidate = exports.throwInvalidAssert = exports.createUnionTypeErrorIterable = exports.createUnionTypeError = exports.createErrorsIterable = exports.adjustErrorMessage = exports.DEFAULT_UNKNOWN_TYPE_MESSAGE = exports.DEFAULT_OVERALL_MESSAGE = void 0;
const typebox_1 = require("@sinclair/typebox");
const errors_1 = require("@sinclair/typebox/errors");
const validation_exception_1 = require("./validation-exception");
exports.DEFAULT_OVERALL_MESSAGE = 'Invalid value';
exports.DEFAULT_UNKNOWN_TYPE_MESSAGE = 'Object type not recognized';
const TYPEBOX_REQUIRED_ERROR_MESSAGE = 'Expected required property';
function adjustErrorMessage(error) {
    if (error.schema.errorMessage !== undefined) {
        error.message = error.schema.errorMessage;
    }
    return error;
}
exports.adjustErrorMessage = adjustErrorMessage;
function createErrorsIterable(typeboxErrorIterator) {
    return {
        [Symbol.iterator]: function* () {
            const errors = typeboxErrorIterator[Symbol.iterator]();
            let result = errors.next();
            let customErrorPath = '???'; // signals no prior path ('' can be root path)
            while (result.value !== undefined) {
                const error = result.value;
                const standardMessage = error.message;
                if (error.path !== customErrorPath) {
                    adjustErrorMessage(error);
                    if (error.message != standardMessage) {
                        customErrorPath = error.path;
                        yield error;
                    }
                    else if (
                    // drop 'required' errors for values that have constraints
                    error.message != TYPEBOX_REQUIRED_ERROR_MESSAGE ||
                        ['Any', 'Unknown'].includes(error.schema[typebox_1.Kind])) {
                        yield error;
                    }
                }
                result = errors.next();
            }
        },
    };
}
exports.createErrorsIterable = createErrorsIterable;
function createUnionTypeError(unionSchema, value) {
    var _a;
    return {
        type: errors_1.ValueErrorType.Union,
        path: '',
        schema: unionSchema,
        value,
        message: (_a = unionSchema.errorMessage) !== null && _a !== void 0 ? _a : exports.DEFAULT_UNKNOWN_TYPE_MESSAGE,
    };
}
exports.createUnionTypeError = createUnionTypeError;
function createUnionTypeErrorIterable(typeError) {
    return {
        [Symbol.iterator]: function* () {
            yield typeError;
        },
    };
}
exports.createUnionTypeErrorIterable = createUnionTypeErrorIterable;
function throwInvalidAssert(overallError, firstError) {
    adjustErrorMessage(firstError);
    throw new validation_exception_1.ValidationException(overallError === undefined
        ? exports.DEFAULT_OVERALL_MESSAGE
        : overallError.replace('{error}', validation_exception_1.ValidationException.errorToString(firstError)), [firstError]);
}
exports.throwInvalidAssert = throwInvalidAssert;
function throwInvalidValidate(overallError, errorOrErrors) {
    throw new validation_exception_1.ValidationException(overallError !== null && overallError !== void 0 ? overallError : exports.DEFAULT_OVERALL_MESSAGE, errorOrErrors instanceof errors_1.ValueErrorIterator
        ? [...createErrorsIterable(errorOrErrors)]
        : [errorOrErrors]);
}
exports.throwInvalidValidate = throwInvalidValidate;
