// typebox-validators@0.3.5/abstract/abstract-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractValidator = void 0;
const value_1 = require("@sinclair/typebox/value");
const error_utils_1 = require("../lib/error-utils");
/**
 * Abstract base class for validators, providing validation services for a
 * JSON schema, offering both short-circuting and full validation, supporting
 * custom error messages, and optionally removing unrecognized properties.
 *
 * As of TypeBox version 0.28.13, when validating any given value, TypeBox
 * checks the value's `maxItems` and `maxLength` constraints before testing
 * any other constraints. An API endpoint can therefore protect itself from
 * faulty and malicious clients by short-circuting validation after the
 * first validation error. The `test` method does this, but it does not
 * return any errors. If you want to short-circuit validation and return the
 * first error, use one of the `assert` methods. These methods protect the
 * server from running regex checks on excessively long strings and from
 * running regex checks on all the items of excessively long arrays.
 *
 * @typeParam S Type for a JSON schema, expressed as a TypeBox type.
 */
class AbstractValidator {
    /**
     * @param schema JSON schema against which to validate values. When a schema
     *  provides an `errorMessage` string option, all errors occurring for that
     *  schema (but not for nested schemas) collapse into a single error having
     *  this message. The `errorMessage` option allows you to provide a custom
     *  error message for a schema. For example, an `errorMessage` on a schema
     *  for a property of an object replaces TypeBox's built-in error messages
     *  for errors that occur on that property.
     */
    constructor(schema) {
        this.schema = schema;
    }
    /**
     * Tests whether a value conforms to the schema, returning an iterable whose
     * iterator yields the validation errors, or returning `null` if there are no
     * validation errors. This method is equivalent to calling `test()` and then
     * `errors()` and exists only for convenience. The method does not throw
     * `ValidationException` and does not clean values of unrecognized properties.
     *
     * @param value Value to validate against the schema.
     * @returns An iteratable yielding all validation errors, if any, otherwise
     *  `null`. Upon detecting one or more errors for a particular schema
     *  (possibly a nested schema), if the schema provides an `errorMessage`
     *  property, only a single error is reported for the schema, and the
     *  `message` property of this error is set to `errorMessage`'s value. Also,
     *  the TypeBox error "Expected required property" is dropped when at least
     *  one other error is reported for the property. Consequently, only the
     *  `Type.Any` and `Type.Unknown` schemas can yield "Expected required
     *  property" errors.
     */
    testReturningErrors(value) {
        return this.test(value) ? null : this.errors(value);
    }
    /**
     * Tests whether a value conforms to the schema, returning the first error,
     * or returning `null` if there is no error. This method is equivalent to
     * calling `test()` and then `firstError()` and exists only for convenience.
     * The method does not throw `ValidationException` and does not clean values
     * of unrecognized properties.
     *
     * @param value Value to validate against the schema.
     * @returns The first validation error, if there is a validation error,
     *  otherwise `null`.
     */
    testReturningFirstError(value) {
        const iterable = this.testReturningErrors(value);
        if (iterable === null) {
            return null;
        }
        const result = iterable[Symbol.iterator]().next();
        return result.done ? null : result.value;
    }
    /**
     * Validates a value against the schema and returns the first error,
     * returning `null` if there is no error. No validation is performed beyond
     * the first error, allowing you to protect the server from wasting time and
     * memory validating excessively long strings. It is equivalent to calling
     * `next()` exactly once on the iterator returned by `errors()`, serving
     * only as a convenience method. For performance reasons, it is best to call
     * `test()` before calling this method. This method does not throw
     * `ValidationException` and does not clean values of unrecognized properties.
     *
     * @param value Value to validate against the schema.
     * @returns The first validation error, if there is a validation error,
     *  otherwise `null`.
     */
    firstError(value) {
        const iterator = this.errors(value)[Symbol.iterator]();
        const result = iterator.next();
        return result.done ? null : result.value;
    }
    cleanCopyOfValue(schema, value) {
        if (schema.type === 'object' && typeof value === 'object') {
            const cleanedValue = {};
            Object.keys(schema.properties).forEach((key) => {
                cleanedValue[key] = value[key];
            });
            return cleanedValue;
        }
        return value;
    }
    cleanValue(schema, value) {
        if (schema.type === 'object' && typeof value === 'object') {
            const schemaKeys = Object.keys(schema.properties);
            Object.getOwnPropertyNames(value).forEach((key) => {
                if (!schemaKeys.includes(key)) {
                    delete value[key];
                }
            });
        }
    }
    uncompiledAssert(schema, value, overallError) {
        if (!value_1.Value.Check(schema, value)) {
            (0, error_utils_1.throwInvalidAssert)(overallError, value_1.Value.Errors(schema, value).First());
        }
    }
    uncompiledValidate(schema, value, overallError) {
        if (!value_1.Value.Check(schema, value)) {
            (0, error_utils_1.throwInvalidValidate)(overallError, value_1.Value.Errors(schema, value));
        }
    }
}
exports.AbstractValidator = AbstractValidator;
