// typebox-validators@0.3.5/abstract/abstract-typed-union-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractTypedUnionValidator = exports.DEFAULT_DISCRIMINANT_KEY = void 0;
const abstract_validator_1 = require("./abstract-validator");
/**
 * The key providing the object type in discriminated unions, if not
 * specified in the schema's `discriminantKey` option.
 */
exports.DEFAULT_DISCRIMINANT_KEY = 'kind';
/**
 * Abstract validator for values that are typed member unions of objects.
 */
class AbstractTypedUnionValidator extends abstract_validator_1.AbstractValidator {
    constructor(schema) {
        super(schema);
    }
    /** @inheritdoc */
    assert(value, overallError) {
        this.assertReturningSchema(value, overallError);
    }
    /** @inheritdoc */
    assertAndClean(value, overallError) {
        const schema = this.assertReturningSchema(value, overallError);
        this.cleanValue(schema, value);
    }
    /** @inheritdoc */
    assertAndCleanCopy(value, overallError) {
        const schema = this.assertReturningSchema(value, overallError);
        return this.cleanCopyOfValue(schema, value);
    }
    /** @inheritdoc */
    validate(value, overallError) {
        this.validateReturningSchema(value, overallError);
    }
    /** @inheritdoc */
    validateAndClean(value, overallError) {
        const schema = this.validateReturningSchema(value, overallError);
        this.cleanValue(schema, value);
    }
    /** @inheritdoc */
    validateAndCleanCopy(value, overallError) {
        const schema = this.validateReturningSchema(value, overallError);
        return this.cleanCopyOfValue(schema, value);
    }
    toValueKeyDereference(key) {
        return /^[a-zA-Z_$][a-zA-Z_$0-9]*$/.test(key)
            ? `value.${key}`
            : `value['${key.replace(/'/g, "\\'")}']`;
    }
}
exports.AbstractTypedUnionValidator = AbstractTypedUnionValidator;
