// typebox-validators@0.3.5/abstract/abstract-compiling-typed-union-validator.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractCompilingTypedUnionValidator = void 0;
const value_1 = require("@sinclair/typebox/value");
const compiler_1 = require("@sinclair/typebox/compiler");
const abstract_typed_union_validator_1 = require("./abstract-typed-union-validator");
const error_utils_1 = require("../lib/error-utils");
/**
 * Abstract validatory for typed unions, performing lazy compilation.
 */
class AbstractCompilingTypedUnionValidator extends abstract_typed_union_validator_1.AbstractTypedUnionValidator {
    /** @inheritdoc */
    constructor(schema) {
        super(schema);
        _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests.set(this, void 0);
        __classPrivateFieldSet(this, _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests, new Array(schema.anyOf.length), "f");
    }
    /** @inheritdoc */
    test(value) {
        const memberIndex = this.compiledFindSchemaMemberIndex(value);
        return this.compiledSchemaMemberTest(memberIndex, value);
    }
    /** @inheritdoc */
    errors(value) {
        const indexOrError = this.compiledFindSchemaMemberIndexOrError(value);
        if (typeof indexOrError !== 'number') {
            return (0, error_utils_1.createUnionTypeErrorIterable)(indexOrError);
        }
        return (0, error_utils_1.createErrorsIterable)(value_1.Value.Errors(this.schema.anyOf[indexOrError], value));
    }
    assertReturningSchema(value, overallError) {
        const indexOrError = this.compiledFindSchemaMemberIndexOrError(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidAssert)(overallError, indexOrError);
        }
        const memberSchema = this.schema.anyOf[indexOrError];
        if (!this.compiledSchemaMemberTest(indexOrError, value)) {
            (0, error_utils_1.throwInvalidAssert)(overallError, value_1.Value.Errors(memberSchema, value).First());
        }
        return memberSchema;
    }
    validateReturningSchema(value, overallError) {
        const indexOrError = this.compiledFindSchemaMemberIndexOrError(value);
        if (typeof indexOrError !== 'number') {
            (0, error_utils_1.throwInvalidValidate)(overallError, indexOrError);
        }
        const memberSchema = this.schema.anyOf[indexOrError];
        if (!this.compiledSchemaMemberTest(indexOrError, value)) {
            (0, error_utils_1.throwInvalidValidate)(overallError, value_1.Value.Errors(memberSchema, value));
        }
        return memberSchema;
    }
    compiledFindSchemaMemberIndexOrError(value) {
        const memberIndex = this.compiledFindSchemaMemberIndex(value);
        if (memberIndex === null) {
            return (0, error_utils_1.createUnionTypeError)(this.schema, value);
        }
        return memberIndex;
    }
    compiledSchemaMemberTest(memberIndex, value) {
        if (memberIndex === null) {
            return false;
        }
        if (__classPrivateFieldGet(this, _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests, "f")[memberIndex] === undefined) {
            let code = compiler_1.TypeCompiler.Compile(this.schema.anyOf[memberIndex]).Code();
            code = code.replace(`(typeof value === 'object' && value !== null && !Array.isArray(value)) &&`, '');
            // provide some resilience to change in TypeBox compiled code formatting
            const startOfFunction = code.indexOf('function');
            const startOfReturn = code.indexOf('return', startOfFunction);
            code =
                'return ' +
                    code.substring(code.indexOf('(', startOfReturn), code.length - 1);
            __classPrivateFieldGet(this, _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests, "f")[memberIndex] = new Function('value', code);
        }
        return __classPrivateFieldGet(this, _AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests, "f")[memberIndex](value);
    }
}
exports.AbstractCompilingTypedUnionValidator = AbstractCompilingTypedUnionValidator;
_AbstractCompilingTypedUnionValidator_compiledSchemaMemberTests = new WeakMap();
