// typebox-validators@0.3.5/standard/compiling-standard-validator.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _CompilingStandardValidator_compiledType;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompilingStandardValidator = void 0;
const compiler_1 = require("@sinclair/typebox/compiler");
const abstract_standard_validator_1 = require("../abstract/abstract-standard-validator");
const error_utils_1 = require("../lib/error-utils");
/**
 * Lazily compiled validator for standard TypeBox values.
 */
class CompilingStandardValidator extends abstract_standard_validator_1.AbstractStandardValidator {
    /** @inheritdoc */
    constructor(schema) {
        super(schema);
        _CompilingStandardValidator_compiledType.set(this, void 0);
    }
    /** @inheritdoc */
    test(value) {
        const compiledType = this.getCompiledType();
        return compiledType.Check(value);
    }
    /** @inheritdoc */
    assert(value, overallError) {
        const compiledType = this.getCompiledType();
        if (!compiledType.Check(value)) {
            (0, error_utils_1.throwInvalidAssert)(overallError, compiledType.Errors(value).First());
        }
    }
    /** @inheritdoc */
    validate(value, overallError) {
        const compiledType = this.getCompiledType();
        if (!compiledType.Check(value)) {
            (0, error_utils_1.throwInvalidValidate)(overallError, compiledType.Errors(value));
        }
    }
    /** @inheritdoc */
    errors(value) {
        const compiledType = this.getCompiledType();
        return (0, error_utils_1.createErrorsIterable)(compiledType.Errors(value));
    }
    getCompiledType() {
        if (__classPrivateFieldGet(this, _CompilingStandardValidator_compiledType, "f") === undefined) {
            __classPrivateFieldSet(this, _CompilingStandardValidator_compiledType, compiler_1.TypeCompiler.Compile(this.schema), "f");
        }
        return __classPrivateFieldGet(this, _CompilingStandardValidator_compiledType, "f");
    }
}
exports.CompilingStandardValidator = CompilingStandardValidator;
_CompilingStandardValidator_compiledType = new WeakMap();
